// @taiga-ui/kit/esm2022/components/select/select-option/select-option.component.mjs
import { ChangeDetectionStrategy, Component, computed, inject } from '@angular/core';
import { TuiControl } from '@taiga-ui/cdk/classes';
import { tuiDirectiveBinding, tuiIsPresent } from '@taiga-ui/cdk/utils/miscellaneous';
import { TUI_DATA_LIST_HOST, TuiOptionWithValue, } from '@taiga-ui/core/components/data-list';
import { TuiScrollIntoView } from '@taiga-ui/core/components/scrollbar';
import { tuiInjectAuxiliary } from '@taiga-ui/core/components/textfield';
import { TuiIcons } from '@taiga-ui/core/directives/icons';
import { TUI_ITEMS_HANDLERS } from '@taiga-ui/core/directives/items-handlers';
import { TUI_COMMON_ICONS } from '@taiga-ui/core/tokens';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/components/scrollbar";
class TuiSelectOption {
    constructor() {
        this.host = inject(TUI_DATA_LIST_HOST);
        this.itemsHandlers = inject(TUI_ITEMS_HANDLERS);
        this.control = tuiInjectAuxiliary((x) => x instanceof TuiControl);
        this.option = inject(TuiOptionWithValue, {
            optional: true,
        });
        this.icon = this.option &&
            tuiDirectiveBinding(TuiIcons, 'iconEnd', inject(TUI_COMMON_ICONS).check, {});
        this.selected = computed((controlValue = this.control()?.value(), optionValue = this.option?.value()) => tuiIsPresent(optionValue) &&
            tuiIsPresent(controlValue) &&
            this.itemsHandlers.identityMatcher()(controlValue, optionValue));
        this.scrolled = tuiDirectiveBinding(TuiScrollIntoView, 'tuiScrollIntoView', this.selected);
    }
    onClick() {
        const value = this.option?.value();
        if (this.host?.handleOption && value !== undefined) {
            this.host.handleOption(value);
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSelectOption, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: TuiSelectOption, isStandalone: true, selector: "ng-component", host: { listeners: { "click": "onClick()" }, properties: { "class._value": "option", "class._selected": "selected()" } }, hostDirectives: [{ directive: i1.TuiScrollIntoView }], ngImport: i0, template: '', isInline: true, styles: [":host._value:after{color:var(--tui-text-action);opacity:0}:host-context(tui-data-list[data-size=\"s\"]) :host._value:after{margin-inline-end:-.625rem}:host-context(tui-data-list[data-size=\"m\"]) :host._value:after,:host-context(tui-data-list[data-size=\"l\"]) :host._value:after{margin-inline-end:-.5rem}:host-context(tui-data-list[data-size=\"l\"]) :host._value:after{font-size:1.5rem}:host._value._selected:after{opacity:1}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
export { TuiSelectOption };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSelectOption, decorators: [{
            type: Component,
            args: [{ standalone: true, template: '', changeDetection: ChangeDetectionStrategy.OnPush, hostDirectives: [TuiScrollIntoView], host: {
                        '[class._value]': 'option',
                        '[class._selected]': 'selected()',
                        '(click)': 'onClick()',
                    }, styles: [":host._value:after{color:var(--tui-text-action);opacity:0}:host-context(tui-data-list[data-size=\"s\"]) :host._value:after{margin-inline-end:-.625rem}:host-context(tui-data-list[data-size=\"m\"]) :host._value:after,:host-context(tui-data-list[data-size=\"l\"]) :host._value:after{margin-inline-end:-.5rem}:host-context(tui-data-list[data-size=\"l\"]) :host._value:after{font-size:1.5rem}:host._value._selected:after{opacity:1}\n"] }]
        }] });
