// @taiga-ui/legacy/esm2022/components/tooltip/tooltip.module.mjs
import { NgModule } from '@angular/core';
import { TuiIcon } from '@taiga-ui/core/components/icon';
import { TuiTooltip } from '@taiga-ui/kit/directives/tooltip';
import { TuiTooltipComponent } from './tooltip.component';
import * as i0 from "@angular/core";
/**
 * @deprecated: drop in v5.0 use {@link TuiTooltip}
 * https://taiga-ui.dev/components/tooltip
 */
class TuiTooltipModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTooltipModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiTooltipModule, declarations: [TuiTooltipComponent], imports: [TuiIcon, TuiTooltip], exports: [TuiTooltipComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTooltipModule, imports: [TuiIcon] }); }
}
export { TuiTooltipModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiTooltipModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [TuiIcon, TuiTooltip],
                    declarations: [TuiTooltipComponent],
                    exports: [TuiTooltipComponent],
                }]
        }] });
