// @taiga-ui/addon-doc/esm2022/components/internal/source-code/source-code.component.mjs
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, inject, Input } from '@angular/core';
import { TUI_DOC_ICONS, TUI_DOC_SOURCE_CODE, TUI_DOC_SOURCE_CODE_TEXT, } from '@taiga-ui/addon-doc/tokens';
import { tuiPure } from '@taiga-ui/cdk/utils/miscellaneous';
import { TuiButton } from '@taiga-ui/core/components/button';
import { PolymorpheusOutlet, PolymorpheusTemplate } from '@taiga-ui/polymorpheus';
import * as i0 from "@angular/core";
class TuiDocSourceCode {
    constructor() {
        this.icons = inject(TUI_DOC_ICONS);
        this.sourceCode = inject(TUI_DOC_SOURCE_CODE);
        this.text = inject(TUI_DOC_SOURCE_CODE_TEXT);
        this.header = '';
        this.package = '';
        this.type = '';
        this.path = '';
    }
    get pathOptions() {
        return this.getPathOptions(this.header, this.package, this.type, this.path);
    }
    pathIsUrl(path) {
        return path.startsWith('http');
    }
    getPathOptions(header, packageName, type, path) {
        return {
            header,
            package: packageName,
            type,
            path,
        };
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDocSourceCode, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: TuiDocSourceCode, isStandalone: true, selector: "tui-doc-source-code", inputs: { header: "header", package: "package", type: "type", path: "path" }, ngImport: i0, template: "<a\n    *polymorpheusOutlet=\"pathIsUrl(path) ? path : sourceCode as link; context: pathOptions\"\n    appearance=\"icon\"\n    size=\"s\"\n    target=\"_blank\"\n    tuiIconButton\n    type=\"button\"\n    [href]=\"link\"\n    [iconStart]=\"icons.code\"\n    [style.height]=\"'var(--tui-height-l)'\"\n>\n    {{ text }}\n</a>\n", dependencies: [{ kind: "directive", type: PolymorpheusOutlet, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }, { kind: "directive", type: TuiButton, selector: "a[tuiButton],button[tuiButton],a[tuiIconButton],button[tuiIconButton]", inputs: ["size"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
__decorate([
    tuiPure
], TuiDocSourceCode.prototype, "pathIsUrl", null);
__decorate([
    tuiPure
], TuiDocSourceCode.prototype, "getPathOptions", null);
export { TuiDocSourceCode };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDocSourceCode, decorators: [{
            type: Component,
            args: [{ standalone: true, selector: 'tui-doc-source-code', imports: [PolymorpheusOutlet, PolymorpheusTemplate, TuiButton], changeDetection: ChangeDetectionStrategy.OnPush, template: "<a\n    *polymorpheusOutlet=\"pathIsUrl(path) ? path : sourceCode as link; context: pathOptions\"\n    appearance=\"icon\"\n    size=\"s\"\n    target=\"_blank\"\n    tuiIconButton\n    type=\"button\"\n    [href]=\"link\"\n    [iconStart]=\"icons.code\"\n    [style.height]=\"'var(--tui-height-l)'\"\n>\n    {{ text }}\n</a>\n" }]
        }], propDecorators: { header: [{
                type: Input
            }], package: [{
                type: Input
            }], type: [{
                type: Input
            }], path: [{
                type: Input
            }], pathIsUrl: [], getPathOptions: [] } });
