// @taiga-ui/core/esm2022/components/data-list/option/option-content.mjs
import { ContentChild, Directive, inject } from '@angular/core';
import { tuiCreateToken } from '@taiga-ui/cdk/utils/miscellaneous';
import * as i0 from "@angular/core";
/**
 * Content for tuiOption component
 */
export const TUI_OPTION_CONTENT = tuiCreateToken();
export function tuiAsOptionContent(useValue) {
    return {
        provide: TUI_OPTION_CONTENT,
        useValue,
    };
}
class TuiWithOptionContent {
    constructor() {
        this.localContent = null;
        this.globalContent = inject(TUI_OPTION_CONTENT, {
            optional: true,
        });
    }
    get content() {
        return this.globalContent ?? this.localContent;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiWithOptionContent, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiWithOptionContent, isStandalone: true, queries: [{ propertyName: "localContent", first: true, predicate: TUI_OPTION_CONTENT, descendants: true }], ngImport: i0 }); }
}
export { TuiWithOptionContent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiWithOptionContent, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                }]
        }], propDecorators: { localContent: [{
                type: ContentChild,
                args: [TUI_OPTION_CONTENT, { descendants: true }]
            }] } });
