// @taiga-ui/cdk/esm2022/pipes/obfuscate/obfuscate.pipe.mjs
import { inject, Pipe } from '@angular/core';
import { TUI_OBFUSCATE_OPTIONS } from './obfuscate.options';
import * as i0 from "@angular/core";
class TuiObfuscatePipe {
    constructor() {
        this.option = inject(TUI_OBFUSCATE_OPTIONS);
    }
    /**
     * Transforms the input value by obfuscating it according to the specified recipe or symbol.
     *
     * @param value The value to be obfuscated should be a string.
     * @param recipe The name of the recipe to use for obfuscation or a single character symbol for custom obfuscation.
     * @returns The obfuscated string.
     * @throws Will throw an error if the specified recipe is not found.
     */
    transform(value, recipe = '') {
        if (!value) {
            return value;
        }
        if (!recipe) {
            return this.option.default(value);
        }
        const obfuscate = this.option.recipes[recipe];
        if (obfuscate) {
            return obfuscate(value);
        }
        if (recipe.length === 1) {
            return this.option.default(value, recipe[0]);
        }
        const availableRecipes = Object.keys(this.option.recipes).sort();
        throw new Error(`Obfuscate recipe "${recipe}" not found. Available recipes: [${availableRecipes}]`);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiObfuscatePipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiObfuscatePipe, isStandalone: true, name: "tuiObfuscate" }); }
}
export { TuiObfuscatePipe };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiObfuscatePipe, decorators: [{
            type: Pipe,
            args: [{
                    standalone: true,
                    name: 'tuiObfuscate',
                }]
        }] });
