// @taiga-ui/legacy/esm2022/components/combo-box/combo-box-strict.directive.mjs
import { Directive, inject, Input } from '@angular/core';
import { TuiComboBoxComponent } from './combo-box.component';
import * as i0 from "@angular/core";
/**
 * TODO(v5): delete it
 * @deprecated use new version of {@link https://taiga-ui.dev/components/combo-box TuiComboBox} (from @taiga-ui/kit) instead
 */
class TuiComboBoxStrictDirective {
    constructor() {
        this.comboBox = inject((TuiComboBoxComponent));
        this.strict = true;
    }
    onInput() {
        if (!this.strict && this.comboBox.search) {
            this.comboBox.value = this.comboBox.search;
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiComboBoxStrictDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiComboBoxStrictDirective, selector: "tui-combo-box[strict]", inputs: { strict: "strict" }, host: { listeners: { "input": "onInput()" } }, ngImport: i0 }); }
}
export { TuiComboBoxStrictDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiComboBoxStrictDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: false,
                    selector: 'tui-combo-box[strict]',
                    host: {
                        '(input)': 'onInput()',
                    },
                }]
        }], propDecorators: { strict: [{
                type: Input
            }] } });
