// @taiga-ui/addon-doc/esm2022/components/page/page-tab.directive.mjs
import { Directive, inject, Input, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
class TuiDocPageTabConnector {
    constructor() {
        this.template = inject((TemplateRef));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDocPageTabConnector, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiDocPageTabConnector, isStandalone: true, selector: "ng-template[pageTab]", inputs: { pageTab: "pageTab" }, ngImport: i0 }); }
}
export { TuiDocPageTabConnector };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDocPageTabConnector, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'ng-template[pageTab]',
                }]
        }], propDecorators: { pageTab: [{
                type: Input
            }] } });
