// @taiga-ui/kit/esm2022/components/files/input-files/input-files-validator.directive.mjs
import { Directive, inject } from '@angular/core';
import { NG_VALIDATORS, Validators } from '@angular/forms';
import { TuiValidator } from '@taiga-ui/cdk/directives/validator';
import { tuiProvide } from '@taiga-ui/cdk/utils/miscellaneous';
import { tuiCreateFileFormatValidator, tuiCreateFileSizeValidator, } from '../files.validators';
import { TUI_INPUT_FILES_OPTIONS } from './input-files.options';
import * as i0 from "@angular/core";
class TuiInputFilesValidator extends TuiValidator {
    constructor() {
        super(...arguments);
        this.options = inject(TUI_INPUT_FILES_OPTIONS);
        this.accept = this.options.accept;
        this.maxFileSize = this.options.maxFileSize;
    }
    ngOnChanges() {
        this.update();
    }
    ngOnInit() {
        this.update();
    }
    update() {
        this.validate =
            Validators.compose([
                tuiCreateFileFormatValidator(this.accept),
                tuiCreateFileSizeValidator(this.maxFileSize),
            ]) || Validators.nullValidator;
        this.onChange();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputFilesValidator, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiInputFilesValidator, isStandalone: true, inputs: { accept: "accept", maxFileSize: "maxFileSize" }, host: { properties: { "accept": "accept" } }, providers: [tuiProvide(NG_VALIDATORS, TuiInputFilesValidator, true)], exportAs: ["tuiInputFilesValidator"], usesInheritance: true, usesOnChanges: true, ngImport: i0 }); }
}
export { TuiInputFilesValidator };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputFilesValidator, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    inputs: ['accept', 'maxFileSize'],
                    providers: [tuiProvide(NG_VALIDATORS, TuiInputFilesValidator, true)],
                    exportAs: 'tuiInputFilesValidator',
                    host: {
                        '[accept]': 'accept',
                    },
                }]
        }] });
