// @taiga-ui/legacy/esm2022/components/multi-select/multi-select-group/multi-select-group.directive.mjs
import { Directive, inject } from '@angular/core';
import { NG_VALUE_ACCESSOR, NgControl } from '@angular/forms';
import { EMPTY_FUNCTION } from '@taiga-ui/cdk/constants';
import { tuiArrayToggle } from '@taiga-ui/cdk/utils/miscellaneous';
import { TUI_DATA_LIST_HOST, tuiAsOptionContent, } from '@taiga-ui/core/components/data-list';
import { TUI_ITEMS_HANDLERS } from '@taiga-ui/kit/tokens';
import { AbstractTuiControl } from '@taiga-ui/legacy/classes';
import { TuiMultiSelectOptionComponent } from '@taiga-ui/legacy/components/multi-select-option';
import { PolymorpheusComponent } from '@taiga-ui/polymorpheus';
import { TuiMultiSelectComponent } from '../multi-select.component';
import * as i0 from "@angular/core";
export const TUI_MULTI_SELECT_OPTION = new PolymorpheusComponent(TuiMultiSelectOptionComponent);
class TuiMultiSelectGroupDirective {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiMultiSelectGroupDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiMultiSelectGroupDirective, selector: "[tuiMultiSelectGroup]", providers: [
            tuiAsOptionContent(TUI_MULTI_SELECT_OPTION),
            {
                provide: TUI_DATA_LIST_HOST,
                useFactory: () => {
                    const multiSelect = inject(TuiMultiSelectComponent, { optional: true });
                    const ngControl = inject(NgControl);
                    const host = inject(AbstractTuiControl, { optional: true });
                    const { identityMatcher, stringify } = inject(TUI_ITEMS_HANDLERS, { optional: true }) ?? {};
                    return (multiSelect || {
                        stringify,
                        identityMatcher,
                        handleOption: (option) => {
                            const value = host?.value ?? ngControl.control?.value ?? [];
                            const result = tuiArrayToggle(value, option, identityMatcher);
                            if (host) {
                                host.value = result;
                            }
                            else {
                                ngControl.control?.setValue(result);
                            }
                        },
                    });
                },
            },
            {
                provide: NG_VALUE_ACCESSOR,
                multi: true,
                useValue: {
                    writeValue: EMPTY_FUNCTION,
                    registerOnChange: EMPTY_FUNCTION,
                    registerOnTouched: EMPTY_FUNCTION,
                },
            },
        ], ngImport: i0 }); }
}
export { TuiMultiSelectGroupDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiMultiSelectGroupDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: false,
                    selector: '[tuiMultiSelectGroup]',
                    providers: [
                        tuiAsOptionContent(TUI_MULTI_SELECT_OPTION),
                        {
                            provide: TUI_DATA_LIST_HOST,
                            useFactory: () => {
                                const multiSelect = inject(TuiMultiSelectComponent, { optional: true });
                                const ngControl = inject(NgControl);
                                const host = inject(AbstractTuiControl, { optional: true });
                                const { identityMatcher, stringify } = inject(TUI_ITEMS_HANDLERS, { optional: true }) ?? {};
                                return (multiSelect || {
                                    stringify,
                                    identityMatcher,
                                    handleOption: (option) => {
                                        const value = host?.value ?? ngControl.control?.value ?? [];
                                        const result = tuiArrayToggle(value, option, identityMatcher);
                                        if (host) {
                                            host.value = result;
                                        }
                                        else {
                                            ngControl.control?.setValue(result);
                                        }
                                    },
                                });
                            },
                        },
                        {
                            provide: NG_VALUE_ACCESSOR,
                            multi: true,
                            useValue: {
                                writeValue: EMPTY_FUNCTION,
                                registerOnChange: EMPTY_FUNCTION,
                                registerOnTouched: EMPTY_FUNCTION,
                            },
                        },
                    ],
                }]
        }] });
