// @taiga-ui/core/esm2022/components/textfield/select-like.directive.mjs
import { ChangeDetectionStrategy, Component, Directive, inject, ViewEncapsulation, } from '@angular/core';
import { TUI_IS_ANDROID } from '@taiga-ui/cdk/tokens';
import { tuiWithStyles } from '@taiga-ui/cdk/utils/miscellaneous';
import * as i0 from "@angular/core";
class TuiSelectLikeStyles {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSelectLikeStyles, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: TuiSelectLikeStyles, isStandalone: true, selector: "ng-component", host: { classAttribute: "tui-select-like" }, ngImport: i0, template: '', isInline: true, styles: [".t-select-like:not(:read-only){cursor:pointer}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSelectLikeStyles, decorators: [{
            type: Component,
            args: [{ standalone: true, template: '', encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, host: {
                        class: 'tui-select-like',
                    }, styles: [".t-select-like:not(:read-only){cursor:pointer}\n"] }]
        }] });
class TuiSelectLike {
    constructor() {
        this.isAndroid = inject(TUI_IS_ANDROID);
        this.nothing = tuiWithStyles(TuiSelectLikeStyles);
    }
    clear(element) {
        element.value = '';
    }
    prevent(event, element) {
        if (this.isAndroid) {
            event.preventDefault();
            element.focus();
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSelectLike, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiSelectLike, isStandalone: true, host: { attributes: { "inputmode": "none", "spellcheck": "false", "autocomplete": "off" }, listeners: { "beforeinput": "$event.inputType.includes(\"delete\") || $event.preventDefault()", "input.capture": "$event.inputType?.includes(\"delete\") && clear($event.target)", "mousedown": "prevent($event, $event.target)" }, classAttribute: "t-select-like" }, ngImport: i0 }); }
}
export { TuiSelectLike };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiSelectLike, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    host: {
                        class: 't-select-like',
                        inputmode: 'none',
                        spellcheck: 'false',
                        autocomplete: 'off',
                        // Click on cleaner icon does not trigger `beforeinput` event --> handle all kind of deletion in input event
                        '(beforeinput)': '$event.inputType.includes("delete") || $event.preventDefault()',
                        '(input.capture)': '$event.inputType?.includes("delete") && clear($event.target)',
                        // Hide Android text select handle (bubble marker below transparent caret)
                        '(mousedown)': 'prevent($event, $event.target)',
                    },
                }]
        }] });
