// @taiga-ui/addon-doc/esm2022/components/documentation/pipes/inspect.pipe.mjs
import { inject, Pipe, TemplateRef } from '@angular/core';
import { tuiInspectAny } from '@taiga-ui/addon-doc/utils';
import { TUI_IS_E2E } from '@taiga-ui/cdk/tokens';
import * as i0 from "@angular/core";
class TuiInspectPipe {
    constructor() {
        this.isE2E = inject(TUI_IS_E2E);
    }
    transform(value, depth = 2) {
        if (this.isE2E && typeof value === 'function') {
            /**
             * @description:
             * When developing in production mode the webpack bundler minify
             * functions in different ways, then due to which the string content
             * of the function may differ from build to build, which can be to
             * various problems when screenshot testing on e2e.
             */
            return 'λ(x) => y';
        }
        return value instanceof TemplateRef ? 'TemplateRef' : tuiInspectAny(value, depth);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInspectPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiInspectPipe, isStandalone: true, name: "tuiInspectAny" }); }
}
export { TuiInspectPipe };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInspectPipe, decorators: [{
            type: Pipe,
            args: [{
                    standalone: true,
                    name: 'tuiInspectAny',
                }]
        }] });
