// @taiga-ui/addon-doc/esm2022/components/documentation/pipes/primitive-polymorpheus-content.pipe.mjs
import { Pipe } from '@angular/core';
import { tuiIsNumber, tuiIsString } from '@taiga-ui/cdk/utils/miscellaneous';
import * as i0 from "@angular/core";
class TuiIsPrimitivePolymorpheusContentPipe {
    transform(value) {
        return tuiIsString(value) || tuiIsNumber(value);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiIsPrimitivePolymorpheusContentPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiIsPrimitivePolymorpheusContentPipe, isStandalone: true, name: "tuiIsPrimitivePolymorpheusContentPipe" }); }
}
export { TuiIsPrimitivePolymorpheusContentPipe };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiIsPrimitivePolymorpheusContentPipe, decorators: [{
            type: Pipe,
            args: [{
                    standalone: true,
                    name: 'tuiIsPrimitivePolymorpheusContentPipe',
                }]
        }] });
