// @taiga-ui/addon-doc/esm2022/components/documentation/pipes/type-reference.pipe.mjs
import { inject, Pipe } from '@angular/core';
import { TUI_DOC_TYPE_REFERENCE_HANDLER, TUI_DOC_TYPE_REFERENCE_PARSER, } from '@taiga-ui/addon-doc/tokens';
import * as i0 from "@angular/core";
class TuiDocTypeReferencePipe {
    constructor() {
        this.parser = inject(TUI_DOC_TYPE_REFERENCE_PARSER);
        this.linkHandler = inject(TUI_DOC_TYPE_REFERENCE_HANDLER);
    }
    transform(original) {
        return this.parser(original)
            .map(({ type, extracted }) => ({
            type,
            extracted,
            reference: this.linkHandler?.(extracted) ?? null,
        }))
            .sort((a, b) => b.reference?.localeCompare(a.reference ?? '') ?? -1);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDocTypeReferencePipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiDocTypeReferencePipe, isStandalone: true, name: "tuiDocTypeReference" }); }
}
export { TuiDocTypeReferencePipe };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDocTypeReferencePipe, decorators: [{
            type: Pipe,
            args: [{
                    standalone: true,
                    name: 'tuiDocTypeReference',
                }]
        }] });
