// @taiga-ui/legacy/esm2022/components/input-date/input-date.directive.mjs
import { Directive } from '@angular/core';
import { AbstractTuiTextfieldHost } from '@taiga-ui/legacy/classes';
import { tuiAsTextfieldHost } from '@taiga-ui/legacy/tokens';
import * as i0 from "@angular/core";
class TuiInputDateDirective extends AbstractTuiTextfieldHost {
    get value() {
        return this.host.computedValue;
    }
    get max() {
        return this.host.computedMax;
    }
    get min() {
        return this.host.computedMin;
    }
    get format() {
        return this.host.dateFormat;
    }
    onValueChange(value) {
        this.host.onValueChange(value);
    }
    process(input) {
        input.inputMode = 'numeric';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputDateDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiInputDateDirective, selector: "tui-input-date:not([multiple])", providers: [tuiAsTextfieldHost(TuiInputDateDirective)], usesInheritance: true, ngImport: i0 }); }
}
export { TuiInputDateDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputDateDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: false,
                    selector: 'tui-input-date:not([multiple])',
                    providers: [tuiAsTextfieldHost(TuiInputDateDirective)],
                }]
        }] });
