// @taiga-ui/addon-doc/esm2022/components/navigation/navigation.providers.mjs
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { ActivatedRoute, NavigationEnd, Router, Scroll } from '@angular/router';
import { TUI_DOC_PAGES, TUI_DOC_TITLE } from '@taiga-ui/addon-doc/tokens';
import { tuiAutoFocusOptionsProvider } from '@taiga-ui/cdk/directives/auto-focus';
import { tuiCreateToken, tuiIsPresent } from '@taiga-ui/cdk/utils/miscellaneous';
import { tuiLinkOptionsProvider } from '@taiga-ui/core/components/link';
import { tuiScrollbarOptionsProvider } from '@taiga-ui/core/components/scrollbar';
import { filter, map, mergeMap } from 'rxjs';
function labelsProviderFactory(pages) {
    return pages
        .map(({ section }) => section)
        .filter(tuiIsPresent)
        .filter((item, index, array) => array.indexOf(item) === index);
}
/**
 * Page title
 */
export const NAVIGATION_TITLE = tuiCreateToken();
/**
 * Navigation sections labels for search
 */
export const NAVIGATION_LABELS = tuiCreateToken();
/**
 * Navigation pages
 */
export const NAVIGATION_ITEMS = tuiCreateToken();
export const NAVIGATION_PROVIDERS = [
    tuiAutoFocusOptionsProvider({ preventScroll: true }),
    tuiLinkOptionsProvider({ appearance: 'action-grayscale' }),
    {
        provide: NAVIGATION_TITLE,
        deps: [Router, ActivatedRoute, TUI_DOC_TITLE],
        useFactory: (router, activatedRoute, titlePrefix) => router.events.pipe(filter((event) => event instanceof NavigationEnd ||
            (event instanceof Scroll
                ? event.routerEvent instanceof NavigationEnd
                : false)), map(() => activatedRoute.firstChild), filter(tuiIsPresent), mergeMap(({ data }) => data), map(({ title }) => `${titlePrefix}${title}`), takeUntilDestroyed()),
    },
    {
        provide: NAVIGATION_LABELS,
        deps: [TUI_DOC_PAGES],
        useFactory: labelsProviderFactory,
    },
    {
        provide: NAVIGATION_ITEMS,
        deps: [TUI_DOC_PAGES],
        useFactory: (pages) => {
            const labels = labelsProviderFactory(pages);
            return [
                ...labels.map((label) => pages.filter(({ section }) => section === label)),
                pages.filter((page) => !page.section),
            ];
        },
    },
    tuiScrollbarOptionsProvider({
        mode: 'hover',
    }),
];
