// @taiga-ui/addon-doc/esm2022/components/navigation/scroll-into-view.directive.mjs
import { Directive, inject, Input } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { TUI_DOC_PAGE_LOADED } from '@taiga-ui/addon-doc/tokens';
import { tuiZonefreeScheduler } from '@taiga-ui/cdk/observables';
import { tuiGetElementObscures, tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import { debounceTime, filter, ReplaySubject, switchMap, take } from 'rxjs';
import * as i0 from "@angular/core";
class TuiDocScrollIntoViewLink {
    constructor() {
        this.scroll$ = new ReplaySubject(1);
        this.el = tuiInjectElement();
        this.sub = inject(TUI_DOC_PAGE_LOADED)
            .pipe(filter(Boolean), take(1), switchMap(() => this.scroll$), debounceTime(750, tuiZonefreeScheduler()), filter((shallWe) => shallWe && !!tuiGetElementObscures(this.el)), takeUntilDestroyed())
            .subscribe(() => this.el.scrollIntoView());
    }
    set tuiDocScrollIntoViewLink(shallWe) {
        this.scroll$.next(shallWe);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDocScrollIntoViewLink, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiDocScrollIntoViewLink, isStandalone: true, selector: "[tuiDocScrollIntoViewLink]", inputs: { tuiDocScrollIntoViewLink: "tuiDocScrollIntoViewLink" }, ngImport: i0 }); }
}
export { TuiDocScrollIntoViewLink };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDocScrollIntoViewLink, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiDocScrollIntoViewLink]',
                }]
        }], propDecorators: { tuiDocScrollIntoViewLink: [{
                type: Input
            }] } });
