// @taiga-ui/kit/esm2022/components/items-with-more/items-with-more.directive.mjs
import { Directive, Input } from '@angular/core';
import { tuiInjectElement } from '@taiga-ui/cdk/utils/dom';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
class TuiItemsWithMoreDirective {
    constructor() {
        this.el = tuiInjectElement();
        this.itemsLimit = Infinity;
        this.required = -1;
        this.linesLimit = 1;
        this.side = 'end';
        // TODO: refactor to signal inputs after Angular update
        this.change$ = new Subject();
    }
    get computedSide() {
        return this.linesLimit > 1 ? 'end' : this.side;
    }
    ngOnChanges() {
        this.change$.next();
    }
    maxWidth() {
        return Math.max(...Array.from(this.el.children, ({ clientWidth }) => clientWidth));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiItemsWithMoreDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiItemsWithMoreDirective, isStandalone: true, inputs: { itemsLimit: "itemsLimit", required: "required", linesLimit: "linesLimit", side: "side" }, host: { properties: { "class._multiline": "linesLimit > 1", "style.--t-min-width.px": "maxWidth()" } }, usesOnChanges: true, ngImport: i0 }); }
}
export { TuiItemsWithMoreDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiItemsWithMoreDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    host: {
                        '[class._multiline]': 'linesLimit > 1',
                        '[style.--t-min-width.px]': 'maxWidth()',
                    },
                }]
        }], propDecorators: { itemsLimit: [{
                type: Input
            }], required: [{
                type: Input
            }], linesLimit: [{
                type: Input
            }], side: [{
                type: Input
            }] } });
