// @taiga-ui/cdk/esm2022/pipes/animation/animation.pipe.mjs
import { inject, Pipe } from '@angular/core';
import { TUI_ANIMATIONS_SPEED } from '@taiga-ui/core/tokens';
import { tuiGetDuration } from '@taiga-ui/core/utils/miscellaneous';
import * as i0 from "@angular/core";
class TuiAnimationPipe {
    constructor() {
        this.speed = inject(TUI_ANIMATIONS_SPEED);
        this.transform = ({ duration, speed, value, ...params } = {}) => ({
            value: value ?? '',
            params: {
                ...params,
                duration: (duration ?? 1) * tuiGetDuration(speed ?? this.speed),
            },
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiAnimationPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TuiAnimationPipe, isStandalone: true, name: "tuiAnimation" }); }
}
export { TuiAnimationPipe };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiAnimationPipe, decorators: [{
            type: Pipe,
            args: [{
                    standalone: true,
                    name: 'tuiAnimation',
                }]
        }] });
