// @taiga-ui/legacy/esm2022/components/input-date-time/native-date-time/native-date-time.directive.mjs
import { Directive, inject } from '@angular/core';
import { TuiDay, TuiTime } from '@taiga-ui/cdk/date-time';
import { TuiInputDateTimeDirective } from '../input-date-time.directive';
import * as i0 from "@angular/core";
class TuiNativeDateTimeDirective {
    constructor() {
        this.host = inject(TuiInputDateTimeDirective);
    }
    get value() {
        if (!this.host.rawValue?.[0] || !this.host.rawValue?.[1]) {
            return '';
        }
        return `${this.host.rawValue[0]?.toString('YMD', '-')}T${this.host.rawValue[1]?.toString('HH:MM')}`;
    }
    onChange(value) {
        if (!value) {
            this.host.writeValue(null);
            return;
        }
        const date = new Date(value);
        const day = TuiDay.fromLocalNativeDate(date);
        const time = TuiTime.fromLocalNativeDate(date);
        this.host.writeValue([day, time]);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiNativeDateTimeDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiNativeDateTimeDirective, selector: "input[tuiDateTime]", host: { attributes: { "type": "datetime-local" }, listeners: { "change.stop": "onChange($event.target.value)", "input.stop": "onChange($event.target.value)", "click.stop.zoneless": "0", "mousedown.stop.zoneless": "0" }, properties: { "tabIndex": "-1", "value": "value" } }, ngImport: i0 }); }
}
export { TuiNativeDateTimeDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiNativeDateTimeDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: false,
                    selector: 'input[tuiDateTime]',
                    host: {
                        type: 'datetime-local',
                        '[tabIndex]': '-1',
                        '[value]': 'value',
                        '(change.stop)': 'onChange($event.target.value)',
                        '(input.stop)': 'onChange($event.target.value)',
                        '(click.stop.zoneless)': '0',
                        '(mousedown.stop.zoneless)': '0',
                    },
                }]
        }] });
