// @taiga-ui/addon-doc/esm2022/components/api/api-item-number.directive.mjs
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
class TuiDocAPINumberItem {
    constructor() {
        this.min = null;
        this.max = null;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDocAPINumberItem, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiDocAPINumberItem, isStandalone: true, selector: "tr[tuiDocAPIItem][type=number]", inputs: { min: "min", max: "max" }, ngImport: i0 }); }
}
export { TuiDocAPINumberItem };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDocAPINumberItem, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'tr[tuiDocAPIItem][type=number]',
                }]
        }], propDecorators: { min: [{
                type: Input
            }], max: [{
                type: Input
            }] } });
