// @taiga-ui/kit/esm2022/components/input-date/input-date.validator.mjs
import { Directive, effect, inject } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { TUI_ALLOW_SIGNAL_WRITES } from '@taiga-ui/cdk/constants';
import { TuiValidator } from '@taiga-ui/cdk/directives/validator';
import { tuiProvide } from '@taiga-ui/cdk/utils/miscellaneous';
import { TuiItemsHandlersDirective } from '@taiga-ui/core/directives/items-handlers';
import * as i0 from "@angular/core";
class TuiInputDateValidator extends TuiValidator {
    constructor() {
        super(...arguments);
        this.handlers = inject(TuiItemsHandlersDirective);
        this.update = effect(() => {
            this.handlers.disabledItemHandler();
            this.onChange();
        }, TUI_ALLOW_SIGNAL_WRITES);
        this.validate = ({ value }) => value && this.handlers.disabledItemHandler()(value)
            ? { tuiDisabledDate: value }
            : null;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputDateValidator, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiInputDateValidator, isStandalone: true, providers: [tuiProvide(NG_VALIDATORS, TuiInputDateValidator, true)], usesInheritance: true, ngImport: i0 }); }
}
export { TuiInputDateValidator };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiInputDateValidator, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    providers: [tuiProvide(NG_VALIDATORS, TuiInputDateValidator, true)],
                }]
        }] });
