// @taiga-ui/cdk/esm2022/directives/validator/validator.directive.mjs
import { Directive } from '@angular/core';
import { NG_VALIDATORS, Validators } from '@angular/forms';
import { EMPTY_FUNCTION } from '@taiga-ui/cdk/constants';
import { tuiProvide } from '@taiga-ui/cdk/utils';
import * as i0 from "@angular/core";
class TuiValidator {
    constructor() {
        this.onChange = EMPTY_FUNCTION;
        this.validate = Validators.nullValidator;
    }
    registerOnValidatorChange(onChange) {
        this.onChange = onChange;
    }
    ngOnChanges() {
        this.onChange();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiValidator, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiValidator, isStandalone: true, selector: "[tuiValidator]", inputs: { validate: ["tuiValidator", "validate"] }, providers: [tuiProvide(NG_VALIDATORS, TuiValidator, true)], usesOnChanges: true, ngImport: i0 }); }
}
export { TuiValidator };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiValidator, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiValidator]',
                    inputs: ['validate: tuiValidator'],
                    providers: [tuiProvide(NG_VALIDATORS, TuiValidator, true)],
                }]
        }] });
