// @taiga-ui/addon-mobile/esm2022/directives/dropdown-mobile/dropdown-mobile.directive.mjs
import { Directive, inject, Input } from '@angular/core';
import { TUI_IS_MOBILE } from '@taiga-ui/cdk/tokens';
import { tuiIsHTMLElement } from '@taiga-ui/cdk/utils/dom';
import { TUI_OPTION_CONTENT } from '@taiga-ui/core/components/data-list';
import { TuiTextfieldComponent } from '@taiga-ui/core/components/textfield';
import { TUI_DROPDOWN_COMPONENT } from '@taiga-ui/core/directives/dropdown';
import { TuiDropdownMobileComponent } from './dropdown-mobile.component';
import { TuiOptionMobile } from './option-mobile/option-mobile.component';
import * as i0 from "@angular/core";
class TuiDropdownMobile {
    constructor() {
        this.isMobile = inject(TUI_IS_MOBILE);
        this.tuiDropdownMobile = '';
    }
    onMouseDown(event) {
        if (!this.isMobile ||
            !tuiIsHTMLElement(event.target) ||
            !event.target.matches('input,textarea') ||
            this.tuiDropdownMobile) {
            return;
        }
        event.preventDefault();
        event.target.focus({ preventScroll: true });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDropdownMobile, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiDropdownMobile, isStandalone: true, selector: "[tuiDropdownMobile]", inputs: { tuiDropdownMobile: "tuiDropdownMobile" }, host: { listeners: { "mousedown": "onMouseDown($event)" }, properties: { "style.visibility": "\"visible\"" } }, providers: [
            {
                provide: TUI_DROPDOWN_COMPONENT,
                useFactory: () => inject(TUI_IS_MOBILE)
                    ? TuiDropdownMobileComponent
                    : inject(TUI_DROPDOWN_COMPONENT, { skipSelf: true }),
            },
            {
                provide: TUI_OPTION_CONTENT,
                useFactory: () => inject(TUI_IS_MOBILE) &&
                    // TODO(v5): remove it after deletion of all legacy controls
                    inject(TuiTextfieldComponent, { optional: true })
                    ? TuiOptionMobile
                    : inject(TUI_OPTION_CONTENT, { skipSelf: true, optional: true }),
            },
        ], ngImport: i0 }); }
}
export { TuiDropdownMobile };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiDropdownMobile, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiDropdownMobile]',
                    providers: [
                        {
                            provide: TUI_DROPDOWN_COMPONENT,
                            useFactory: () => inject(TUI_IS_MOBILE)
                                ? TuiDropdownMobileComponent
                                : inject(TUI_DROPDOWN_COMPONENT, { skipSelf: true }),
                        },
                        {
                            provide: TUI_OPTION_CONTENT,
                            useFactory: () => inject(TUI_IS_MOBILE) &&
                                // TODO(v5): remove it after deletion of all legacy controls
                                inject(TuiTextfieldComponent, { optional: true })
                                ? TuiOptionMobile
                                : inject(TUI_OPTION_CONTENT, { skipSelf: true, optional: true }),
                        },
                    ],
                    host: {
                        '[style.visibility]': '"visible"',
                        '(mousedown)': 'onMouseDown($event)',
                    },
                }]
        }], propDecorators: { tuiDropdownMobile: [{
                type: Input
            }] } });
