// @taiga-ui/addon-mobile/esm2022/components/mobile-calendar-dropdown/mobile-calendar-dropdown.directive.mjs
import { ContentChild, Directive, inject } from '@angular/core';
import { TUI_IS_MOBILE } from '@taiga-ui/cdk/tokens';
import { TUI_DROPDOWN_COMPONENT } from '@taiga-ui/core/directives/dropdown';
import { TuiItemsHandlersDirective } from '@taiga-ui/core/directives/items-handlers';
import { TuiInputDateDirective } from '@taiga-ui/kit/components/input-date';
import { TuiMobileCalendarDropdown } from './mobile-calendar-dropdown.component';
import * as i0 from "@angular/core";
// TODO: Rename to TuiMobileCalendarDropdown in v5
class TuiMobileCalendarDropdownNew {
    constructor() {
        this.handlers = inject(TuiItemsHandlersDirective);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiMobileCalendarDropdownNew, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: TuiMobileCalendarDropdownNew, isStandalone: true, selector: "[tuiMobileCalendar]", providers: [
            {
                provide: TUI_DROPDOWN_COMPONENT,
                useFactory: () => inject(TUI_IS_MOBILE)
                    ? TuiMobileCalendarDropdown
                    : inject(TUI_DROPDOWN_COMPONENT, { skipSelf: true }),
            },
        ], queries: [{ propertyName: "date", first: true, predicate: TuiInputDateDirective, descendants: true }], ngImport: i0 }); }
}
export { TuiMobileCalendarDropdownNew };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TuiMobileCalendarDropdownNew, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[tuiMobileCalendar]',
                    providers: [
                        {
                            provide: TUI_DROPDOWN_COMPONENT,
                            useFactory: () => inject(TUI_IS_MOBILE)
                                ? TuiMobileCalendarDropdown
                                : inject(TUI_DROPDOWN_COMPONENT, { skipSelf: true }),
                        },
                    ],
                }]
        }], propDecorators: { date: [{
                type: ContentChild,
                args: [TuiInputDateDirective]
            }] } });
